Fix issue 27162 sort order link incorrect compare

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Previously sortOrder number of links not working as aspect. Lower number mean start first. Higher number will stay at last. Seem issue caused by previous changes. Just correct it again

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#27162: SortOrder is reversed in layout XML

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
After this changes can make order of some links in customer navigation account changed

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

lbajsarowicz reviewed (changes requested):
Nice catch! 
Could you provide very basic unit test for that?

mrtuvn commented:
@lbajsarowicz should i need unit test cover this ? Only one simple changed
Can we not require cover test for such case

mrtuvn commented:
@lbajsarowicz updated changes with test unit

lbajsarowicz reviewed (dismissed):
OK, so that means we have Unit Tests covering that feature :-)

:heavy_check_mark: Thank you for your contribution.

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Alfa reviewed (changes requested):
Hi @mrtuvn.
Your changes work, but an order of all links in the navigation menu was changed

**Actual Result:**
:x: 
![screenshotafter](https://user-images.githubusercontent.com/51679138/77147627-5c09b480-6a96-11ea-8903-a31424ac4f77.png)

**Expected Result:**
:heavy_check_mark: order of links on mainline
![screenshot333](https://user-images.githubusercontent.com/51679138/77147942-244f3c80-6a97-11ea-9553-b82272c9b0ff.png)

Maybe you should sort the links in the right order?
@mrtuvn @lbajsarowicz Could you take a look?

Thanks!

mrtuvn commented:
@engcom-Alfa yes this because sort order defined in layout xml. I believe current the order of My account link is highest value
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/view/frontend/layout/customer_account.xml#L27

mrtuvn commented:
hi guys any news on this

lbajsarowicz reviewed (dismissed):
Code looks fine. We need confirmation from QA team if no side effects were introduced.

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa reviewed (changes requested):
Hi, @mrtuvn 

The sorting order must remain the same as before the fix
You can see the sort order by [link](https://docs.magento.com/m2/ce/user_guide/customers/account-dashboard.html) 

**Actual Result:**  :heavy_multiplication_x: 

![Screenshot from 2020-04-01 15-16-46](https://user-images.githubusercontent.com/51679138/78136595-94ad7480-742c-11ea-90cc-9d977fd5532a.png)

**Expected Result:**  :heavy_check_mark: 

![sort](https://user-images.githubusercontent.com/51679138/78137123-68462800-742d-11ea-83dc-9104b8c8e1f7.png)

@mrtuvn Could you take a look?

Thanks!

mrtuvn commented:
@engcom-Alfa updated links order one again

magento-engcom-team commented:
Hi @krzksz, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa reviewed (changes requested):
Hi @mrtuvn,

`Billing Agreements` should be below the  `Stored Payment Methods` block
(Magento/Paypal/view/frontend/layout/customer_account.xml)
**Actual Result:**
![screenshotafter2](https://user-images.githubusercontent.com/51679138/79111321-f81c9800-7d84-11ea-8bc4-6ab32a99faa8.png)

**Expected Result:**
![screenshotbefore](https://user-images.githubusercontent.com/51679138/79111344-036fc380-7d85-11ea-8c8d-1fb65f7b605c.png)

@mrtuvn Could you take a look?
Thanks!

mrtuvn commented:
@engcom-Alfa  updated follow your suggest in my latest commit

krzksz commented:
@magento run all tests

mrtuvn commented:
Hey guys can we start testing on this ?

engcom-Alfa commented:
Hi @krzksz. Could you review the latest changes? Thanks!

magento-engcom-team commented:
Hi @krzksz, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**
1. Add a child block with `sortOrder` value of `1` in your layout XML, e.g., `/view/frontend/layout/customer_account.xml` in your custom theme:

```
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <referenceBlock name="customer_account_navigation">
            <block class="Magento\Customer\Block\Account\SortLinkInterface" name="customer-account-navigation-logout-link">
                <arguments>
                    <argument name="label" xsi:type="string" translate="true">Logout</argument>
                    <argument name="path" xsi:type="string">account/logout</argument>
                    <argument name="sortOrder" xsi:type="number">1</argument>
                </arguments>
            </block>
        </referenceBlock>
    </body>
</page>
```
2. Flush the layout and full page caches
3. Log in at the front end and navigate to `/customer/account/`

**Before:** :heavy_multiplication_x:   A new **Logout** link located at the end of the list

![before](https://user-images.githubusercontent.com/51679138/79553666-7dfe5300-80a5-11ea-9119-f87636d0ef35.png)

**After:** :heavy_check_mark:    A new **Logout** link has the correct position

![after](https://user-images.githubusercontent.com/51679138/79553951-00871280-80a6-11ea-89ae-833e104c7e3d.png)

_**Note:**_ Other items kept the sort order as previous

engcom-Echo commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

engcom-Echo commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

engcom-Echo commented:
@magento run all tests

mrtuvn commented:
@engcom-Echo can you check status

engcom-Echo commented:
Hi @mrtuvn. Related PR must be approved.

mrtuvn commented:
@engcom-Alfa hi can you check this

magento-engcom-team commented:
Hi @dmytro-ch, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

slavvka reviewed (changes requested):
@mrtuvn please resolve conflicts and re-run tests after upgrade to PHP 7.4

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run tests

mrtuvn commented:
@slavvka @lbajsarowicz can you guys review this

mrtuvn commented:
@magento run all tests

magento-engcom-team commented:
Hi @slavvka, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**
1. Add a child block with `sortOrder` value of `1` in your layout XML, e.g., `/view/frontend/layout/customer_account.xml` in your custom theme:

```
<?xml version="1.0"?>
<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
    <body>
        <referenceBlock name="customer_account_navigation">
            <block class="Magento\Customer\Block\Account\SortLinkInterface" name="customer-account-navigation-logout-link">
                <arguments>
                    <argument name="label" xsi:type="string" translate="true">Logout</argument>
                    <argument name="path" xsi:type="string">account/logout</argument>
                    <argument name="sortOrder" xsi:type="number">1</argument>
                </arguments>
            </block>
        </referenceBlock>
    </body>
</page>
```
2. Flush the layout and full page caches
3. Log in at the front end and navigate to `/customer/account/`

**Before:** :heavy_multiplication_x:   A new **Logout** link located at the end of the list

![screenshot_20](https://user-images.githubusercontent.com/51679138/82537607-f6e44500-9b52-11ea-8d9c-db7eb16eb568.png)

**After:** :heavy_check_mark:    A new **Logout** link has the correct position

![screenshot_21](https://user-images.githubusercontent.com/51679138/82537623-fc418f80-9b52-11ea-9e5a-5d689aa9045f.png)


_**Note:**_ Other items kept the sort order as previous

mrtuvn commented:
@magento run all tests

mrtuvn commented:
squash my commits + check conficts if availables

mrtuvn commented:
Hi any update on this ?

eduard13 commented on app/code/Magento/Customer/view/frontend/layout/customer_account.xml:
```diff
@@ -24,31 +24,31 @@
                             <argument name="label" xsi:type="string" translate="true">My Account</argument>
                             <argument name="path" xsi:type="string">customer/account</argument>
-                            <argument name="sortOrder" xsi:type="number">250</argument>
+                            <argument name="sortOrder" xsi:type="number">0</argument>
```
I'm really wondering, how will we add another item before `My Account` as far as this one has `0`?

mrtuvn replied:
If you used negative sortOrder the link will go first before 0. In this case new link will render before My Account link

eduard13 replied:
😄 In my opinion this isn't something correctly and obviously to have negative values for positions.
I'd suggest using a higher number instead of 0.

mrtuvn commented:
Can we have any proceed on this

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

Commit:
Fix incorrect compare sort order link

Fix static test

Fix fail unit test

Update test file

Reorder links via layout xml

Update links order in other extensions

Update order delimiter link

Update link again for toplinks

Update order link after reviews

Remove whitespace anchor tag

Update default.xml

Replace sortOrder zero
Update customer_account.xml

lbajsarowicz commented:
@magento run all tests

Let's verify PR once again and proceed with it.

mrtuvn commented:
can we have put this on higher priority! I have created this 3 months but not merged yet

mrtuvn commented:
@magento run all tests

engcom-Delta reviewed (changes requested):
@eduard13 , @lbajsarowicz Could you review latest commit?

eduard13 reviewed (approved):
Looks good 👍

eduard13 commented:
But looks like something went wrong with this PR, as all checks are failing.

magento-engcom-team commented:
Hi @eduard13, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

mrtuvn commented:
@magento run all tests

mrtuvn commented:
Hi can you guys run tests again ? I have no idea why this happen. Magento no run tests after commits change like before
@slavvka Is this system work normally ?

eduard13 commented:
Most likely it is related with the related PR, that has a conflicting file there. The conflicts should be fixed before running all the tests.

mrtuvn commented:
cherry picks commit and updated mainline code base to my PR branch

Commit:
used escaper in block

mrtuvn commented:
@ihor-sviziev can you re-run tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev committed:
Merge branch '2.4-develop' into fix-issue-27162-sort-order-link-incorrect-compare

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@mrtuvn looks like issue in related b2b pull request. @lbajsarowicz could you help?

mrtuvn commented:
Should i recreate new PR ?

ihor-sviziev commented:
@mrtuvn you don't need to re-create PR

engcom-Echo commented:
@magento run all tests

mrtuvn commented:
Nice!

ihor-sviziev commented:
@eduard13 could  you review PR again as there were some updates and now tests age passing?

eduard13 reviewed (approved):
The changes are good 👍

magento-engcom-team commented:
Hi @eduard13, thank you for the review.
[ENGCOM-7147](https://jira.corp.magento.com/browse/ENGCOM-7147) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

ihor-sviziev commented:
Looks like this PR was reverted in https://github.com/magento/magento2/commit/a360c06684322189609dd543d09bc31d8d2bdca0
@naydav could you explain why it was reverted?

sdzhepa commented:
Hello @ihor-sviziev 

This PR has been reverted due to a regression bug found during testing of 2.4.1

Short info from the report:
> After merging of https://github.com/magento/magento2/pull/27340 the customer links in customer account menu on storefront are messed up in EE and B2B editions. The reason is that pull request #27340 reversed the sort order comparison, but reversed the sort order numbers in layout only for CE.

### Actual Result - link positions are mixed up in b2b
![image-2020-08-06-09-42-46-580](https://user-images.githubusercontent.com/3409584/90829658-c81e6c00-e305-11ea-8a1c-62f304bda003.png)


### Expected result -  link positions are proper in b2b and ee editions.
![image-2020-08-06-09-43-57-016](https://user-images.githubusercontent.com/3409584/90829719-e2584a00-e305-11ea-90ba-71398f7306c2.png)


### Aditional Comment
> ... please consider to revise issues https://github.com/magento/magento2/issues/27162 it is personal expectations on links to be ordered in another direction, it is not a defect.

mrtuvn commented:
why not update order in other b2b or ee modules. I don't think revert code for fix is the good way
invitations, gift registry, reward ..etc and others no doubt should update

ericmorand commented:
@sdzhepa how can having a reversed sort order in customer navigation totally inconsistent with all other places where sort orders are using be a _not a defect_? This looks like a lazy move to ensure backward compatibility with modules that accepted that situation instead of raising an issue when there maintainers faced it. And considering that the modules you are trying to preserve are modules maintained by the core contributor teams (EE and B2B editions), it looks like a big conflict of interest situation when EE and B2B developers are considered as first-class citizen and CE ones are considered secondary.

There is a bug here. It needs to be fixed. And EE and B2B cores need to be fixed accordingly.

mrtuvn commented:
@ericmorand I'm afraid this can take time and much more effort. Since I think a lot a code depend on this. One you changed on this you must change other place

ericmorand commented:
@mrtuvn , I know. This is easy to fix on our side, but the way this issue is handled doesn't cast a glorious light over Magento 2 core maintainers. There is a bug here, you fixed it, and they refused it because it created issues on their side because _they_ used a buggy implementation in the past instead of fixing it as soon as they noticed it. In other words: they force everybody else in the world to fix that issue in their custom code so that they don't have the burden of fixing it on their side.

It's a lack of respect to their community.

ihor-sviziev commented:
Hi @ericmorand,
I got your point, you're totally right in terms of ce vs ee modules. But let's try to see wider.

Unfortunately such change will break not the only b2b or ee extensions , but a lot of custom extensions developed by other people. 

If this change will be accepted, even to 2.5 (as it breaks at least some extensions - it's backward incompatible change) - all extensions that will be affected this change will need to be fixed. There wouldn't be an easy way to fix it both for new version and not breaking compatibility for earlier versions, so the only way will be preparing separate version of extension specially for 2.5 and above. That's not the thing that everybody expects from Magento with their backward compatibility policy.

ericmorand commented:
@ihor-sviziev , I get your point but why is including a breaking change a problem with this specific issue but not a problem for all the previous releases of Magento 2 that included breaking changes?

Don't look any further than 2.4.1. It comes with breaking changes that force everybody to test their code and potentially change them. I would totally agree with you if you were able to point me to the discussions that validated the breaking changes of 2.4.1. Or all other previous releases that included some breaking changes - and there are many of them.

I'm entitled to my opinion: when core maintainers think it's more comfortable for them to include some breaking changes, they have no remorse including them even in patch releases. When they think it's less comfortable for them - like our current case - they refuse perfectly valid PRs.

I may be wrong of course. But that's how it looks like from that side of the project.

So my proposal is : release 2.4.2 that fixes this issue and create a breaking change. Like you did with 2.4.1.

ihor-sviziev commented:
@ericmorand 
Unfortunately when you’re writing some code to magento core you can’t know for sure how exactly it will affect custom extensions.

For instance in case of this PR it looked to me ok to accept such change, but it broke some magento commerce extensions. That’s basically good that it was catched before release. 

I don’t really know why exactly some backward incompatible changes are getting merged to core and includes into patch versions, maybe that was some security issue behind, maybe some other reason, maybe just that case wasn’t noticed.

Would be good if you could provide some examples, so we as a community maybe could analyze them and write some automated tests in order to prevent similar cases in future.

ericmorand commented:
@ihor-sviziev Magento 2 maintainers create breaking change release deliberately:

https://devdocs.magento.com/guides/v2.4/release-notes/backward-incompatible-changes/index.html

https://devdocs.magento.com/guides/v2.4/release-notes/backward-incompatible-changes/reference.html

Just for the two latest releases. I let you check the previous ones.

So it is not unnoticed. Why is this PRs which also create a breaking change not treated as a core PR that create a breaking change ?

ihor-sviziev commented:
Hi @ericmorand,
From what I see - there only one thing that significantly affected upgrades to 2.4.0 - required ElasticSearch. This decision was done by Magento Arcitects, so that's the question to them why they decided.

Currently I have quite strong feeling that we should not accept such change due to inability to make these changes backward compatible.

Could you contact me in Slack and we'll discuss it more detailed?

hostep commented:
I'm ok with approving this for 2.5.0 but not for 2.4.2.

The first one is a major release and it is expected things to break, the second one is a minor release and even though Magento always seem to ship with small backwards incompatible breaking things in minor releases (especially on the frontend), we should really try to prevent this as much as possible, I already have to spend way too much time on such minor/"easy" upgrades (which are never really easy).

gabrieldagama commented:
@ihor-sviziev @hostep @ericmorand @mrtuvn 

We've discussed this issue at our Public Triage Meeting (look at our community calendar for participate), as suggested by other contributors it makes sense to target such changes to the next minor release. We also need to provide an appropriate fix to EE and B2B repositories.

Thanks for highlight and discuss it.

ihor-sviziev commented:
@mrtuvn could you recreate pr with the same changes?

ericmorand commented:
@gabrieldagama thanks a lot, it's great news.

hostep commented:
@gabrieldagama : just to clarify, by "minor" you mean 2.**X**.0 right?

In my opinion we should speak about major versions in this case, the 2 is just a marketing version without any meaning.

- 2.4.0 translates to 4.0.0
- 2.4.0-p1 translates to 4.0.1
- 2.4.1 translates to 4.1.0
- ...

ihor-sviziev commented:
@hostep yes, for sure, it should be included in 2.5.0
